scripts/test-upgrade.sh

```diff
@@ -22,30 +22,59 @@ BASE_DIR=$( cd "$( dirname ${0} )" && cd ../ && pwd )
 RPCD_DIR="$BASE_DIR/rpcd"
 
+# Create a backup of /etc/openstack_deploy, if one doesn't already exist
+# NOTE: This gets done by openstack-ansible's scripts/run-upgrade.sh script,
+#       but if we wait until then it'll be too late.
+if [[ ! -d "/etc/openstack_deploy.MITAKA" ]]; then
+  cp -a /etc/openstack_deploy /etc/openstack_deploy.LIBERTY
+fi
+
 # TASK #0
 # Bug: https://github.com/rcbops/u-suk-dev/issues/347
 # Issue: Variables files have been reorganized, run a migration script to move
 #        them to the new location.
-"${BASE_DIR}"/scripts/migrate-yaml.py --for-testing-take-new-vars-only \
-    --defaults "${RPCD_DIR}"/etc/openstack_deploy/user_rpco_variables_defaults.yml \
-    --overrides /etc/openstack_deploy/user_extras_variables.yml \
-    --output-file /etc/openstack_deploy/user_rpco_variables_overrides.yml
-"${BASE_DIR}"/scripts/migrate-yaml.py --for-testing-take-new-vars-only \
-  --defaults "${RPCD_DIR}"/etc/openstack_deploy/user_osa_variables_defaults.yml \
-  --overrides /etc/openstack_deploy/user_variables.yml \
-  --output-file /etc/openstack_deploy/user_osa_variables_overrides.yml
-cp -a ${RPCD_DIR}/etc/openstack_deploy/*defaults* /etc/openstack_deploy
-rm /etc/openstack_deploy/user_extras_variables.yml /etc/openstack_deploy/user_variables.yml
+if [[ ! -f "/etc/openstack_deploy/user_rpco_variables_overrides.yml" ]]; then
+  "${BASE_DIR}"/scripts/migrate-yaml.py --for-testing-take-new-vars-only \
+      --defaults "${RPCD_DIR}"/etc/openstack_deploy/user_rpco_variables_defaults.yml \
+      --overrides /etc/openstack_deploy/user_extras_variables.yml \
+      --output-file /etc/openstack_deploy/user_rpco_variables_overrides.yml
+fi
+if [[ ! -f "/etc/openstack_deploy/user_osa_variables_overrides.yml" ]]; then
+  "${BASE_DIR}"/scripts/migrate-yaml.py --for-testing-take-new-vars-only \
+    --defaults "${RPCD_DIR}"/etc/openstack_deploy/user_osa_variables_defaults.yml \
+    --overrides /etc/openstack_deploy/user_variables.yml \
+    --output-file /etc/openstack_deploy/user_osa_variables_overrides.yml
+fi
+# TODO: We need to eventually find a proper way to migrate these secret variables.
+#       For now, we just copy existing secret variables defined previously on the
+#       liberty install. This won't cause any problems since there aren't any new
+#       secret variables in mitaka that are not defined in liberty.
+if [[ -f "/etc/openstack_deploy/user_extras_secrets.yml" ]]; then
+  cp /etc/openstack_deploy/user_extras_secrets.yml \
+     /etc/openstack_deploy/user_rpco_secrets.yml
+  cp -a ${RPCD_DIR}/etc/openstack_deploy/*defaults* /etc/openstack_deploy
+  rm /etc/openstack_deploy/user_extras_variables.yml \
+     /etc/openstack_deploy/user_variables.yml        \
+     /etc/openstack_deploy/user_extras_secrets.yml
+fi
 
 # TASK #1
-# Bug: https://github.com/rcbops/u-suk-dev/issues/293
+# Update Ansible
+if [[ -f "/root/.pip/pip.conf" ]]; then
+  rm /root/.pip/pip.conf
+fi
+cd ${OA_DIR}
+bash scripts/bootstrap-ansible.sh
```

Comment:
I believe thats in the deploy.sh script which doesn't actually get called by test-upgrade.sh
scripts/deploy-rpc-playbooks.sh does get called, but that doesn't perform a bootstrap.

We do end up with 2 bootstraps unfortunately, because one happens as part of the openstack-ansible run-upgrade.sh, but that seems to be unavoidable if we want to remove and recreate the repo servers (Task 4). Or at least we do need to take some actions and performing a bootstrap just seems simpler than (for example) installing a subset of roles and hoping all dependencies etc are met.